### ISSUE TITLE
ClickHouse dialect does not support all join types

### ISSUE THREAD

**Issue by jbradt** [sqlfluff/sqlfluff#3506]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

ClickHouse supports more join types than ANSI SQL, including `semi join` and `anti join`. Their docs show a join clause syntax like
```
SELECT <expr_list>
FROM <left_table>
[GLOBAL] [INNER|LEFT|RIGHT|FULL|CROSS] [OUTER|SEMI|ANTI|ANY|ASOF] JOIN <right_table>
(ON <expr_list>)|(USING <column_list>) ...
```
([Source](https://clickhouse.com/docs/en/sql-reference/statements/select/join/#supported-types-of-join))

However, SQLFluff misinterprets the `semi` in `semi join`, for example, as a table alias. It doesn't seem to support join types beyond typical ones like `left join` or `inner join`.

### Expected Behaviour

The tool should understand all of the join types in the syntax above.

### Observed Behaviour

A join query like this one:

```sql
select
    example1.a,
    b
from example1
    semi join example2 on (example1.a = example2.a);
```

yields these warnings (note how it thinks `semi` is an alias for `example1`, while it's really a join type like `left join` or `inner join`):

```
❯ sqlfluff lint example.sql --dialect clickhouse
== [example.sql] FAIL
L:  19 | P:   5 | L027 | Unqualified reference 'b' found in select with more than
                       | one referenced table/view.
L:  21 | P:   5 | L011 | Implicit/explicit aliasing of table.
L:  21 | P:   5 | L025 | Alias 'semi' is never used in SELECT statement.
L:  21 | P:   5 | L031 | Avoid aliases in from clauses and join conditions.
L:  21 | P:  10 | L051 | Join clauses should be fully qualified.
All Finished 📜 🎉!
```

### How to reproduce

```sql
-- example.sql
select
    example1.a,
    b
from example1
    semi join example2 on (example1.a = example2.a);
```

Command:

```
sqlfluff lint example.sql --dialect clickhouse
```

### Dialect

ClickHouse

### Version

```
❯ sqlfluff --version
sqlfluff, version 1.0.0

❯ python --version
Python 3.7.13
```

### Configuration

Just the stock configuration, no other config files.

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by tunetheweb** [comment]
This should be as simple as overriding `JoinKeywordsGrammar` in `dialect_clickhouse.py` and adding a test case as detailed in [Contributing Dialect Changes wiki](https://github.com/sqlfluff/sqlfluff/wiki/Contributing-Dialect-Changes) if anyone fancies a stab at this.

The sparksql dialect has a fairly detailed one as an example:

https://github.com/sqlfluff/sqlfluff/blob/7d9717dbec193107ae4610d0c9139ef8a0bbeef7/src/sqlfluff/dialects/dialect_sparksql.py#L500-L519

**Comment by konnectr** [comment]
I started developing this issue. This issue is not as simple as it seems at first glance.
